bpf/utils.h

```diff
@@ -0,0 +1,299 @@
+     id->eth_protocol = bpf_ntohs(eth.h_proto);
+     __builtin_memcpy(id->dst_mac, eth.h_dest, ETH_ALEN);
+     __builtin_memcpy(id->src_mac, eth.h_source, ETH_ALEN);
+    if (id->eth_protocol == ETH_P_IP) {
+        *family = AF_INET;
+    } else if (id->eth_protocol == ETH_P_IPV6) {
+        *family = AF_INET6;
+    }
+ }
+
+static inline void set_key_with_l3_info(struct sk_buff *skb, u16 family, flow_id *id, u8 *protocol) {
+     if (family == AF_INET) {
+         struct iphdr ip;
+         __builtin_memset(&ip, 0, sizeof(ip));
+         bpf_probe_read(&ip, sizeof(ip), (struct iphdr *)(skb->head + skb->network_header));
+         __builtin_memcpy(id->src_ip, ip4in6, sizeof(ip4in6));
+         __builtin_memcpy(id->dst_ip, ip4in6, sizeof(ip4in6));
+         __builtin_memcpy(id->src_ip + sizeof(ip4in6), &ip.saddr, sizeof(ip.saddr));
+         __builtin_memcpy(id->dst_ip + sizeof(ip4in6), &ip.daddr, sizeof(ip.daddr));
+         *protocol = ip.protocol;
+     } else if (family == AF_INET6) {
+         struct ipv6hdr ip;
+         __builtin_memset(&ip, 0, sizeof(ip));
+         bpf_probe_read(&ip, sizeof(ip), (struct ipv6hdr *)(skb->head + skb->network_header));
+         __builtin_memcpy(id->src_ip, ip.saddr.in6_u.u6_addr8, IP_MAX_LEN);
+         __builtin_memcpy(id->dst_ip, ip.daddr.in6_u.u6_addr8, IP_MAX_LEN);
+         *protocol = ip.nexthdr;
+     }
+ }
+
+static inline int set_key_with_tcp_info(struct sk_buff *skb, flow_id *id, u8 protocol, u16 *flags) {
+     u16 sport = 0,dport = 0;
+     struct tcphdr tcp;
+
+     __builtin_memset(&tcp, 0, sizeof(tcp));
+     bpf_probe_read(&tcp, sizeof(tcp), (struct tcphdr *)(skb->head + skb->transport_header));
+     sport = bpf_ntohs(tcp.source);
+     dport = bpf_ntohs(tcp.dest);
+     set_flags(&tcp, flags);
+     id->src_port = sport;
+     id->dst_port = dport;
+     id->transport_protocol = protocol;
+     return tcp.doff * sizeof(u32);
+ }
+
+static inline int set_key_with_udp_info(struct sk_buff *skb, flow_id *id, u8 protocol) {
+     u16 sport = 0,dport = 0;
+     struct udphdr udp;
+
+     __builtin_memset(&udp, 0, sizeof(udp));
+     bpf_probe_read(&udp, sizeof(udp), (struct udp *)(skb->head + skb->transport_header));
+     sport = bpf_ntohs(udp.source);
+     dport = bpf_ntohs(udp.dest);
+     id->src_port = sport;
+     id->dst_port = dport;
+     id->transport_protocol = protocol;
+     return bpf_ntohs(udp.len);
+ }
+
+static inline long tcp_drop_lookup_and_update_flow(struct sk_buff *skb, flow_id *id, u8 state, u16 flags,
+                                            enum skb_drop_reason reason) {
+     flow_metrics *aggregate_flow = bpf_map_lookup_elem(&aggregated_flows, id);
+     if (aggregate_flow != NULL) {
+         aggregate_flow->tcp_drops.packets += 1;
+         aggregate_flow->tcp_drops.bytes += skb->len;
+         aggregate_flow->tcp_drops.state = state;
+         aggregate_flow->tcp_drops.flags |= flags;
+         aggregate_flow->tcp_drops.drop_cause = reason;
```

Comment:
Having these variables in an "aggregate flow" entry seems a bit weird. So the `state` and `drop_reason` fields will be set to the last seen socket state and drop cause for the flow, and the flags will be a bitwise OR of all packet flags ever seen on the connection? Just overriding values with the last value seen seems like a good way to lose data; and the bitwise combination of the flags just doesn't make any semantic sense. What are there fields actually meant to be used for?

Reply:
Well, depends. What do you do when you get multiple drops on the same flow (that is quite normal in a TCP session)? I can think of two options:
- Always create a new flow entry for each drop.
- Keep the flow entry, but have multiple drop events associated with each flow

From a protocol perspective, the latter is more correct (logically, the multiple drops would be associated with the same flow), and you could have a drop counter for a flow that would increment, etc. You'll need to rework the data structure so you can accommodate that, though...

Reply:
Well that's the thing that's by no means clear to me (whether the last drop reason is the important one). But it that's the use case, fair enough; definitely get rid of the `|=` on the flags field, though (or just lose the field entirely)...

Reply:
As user I see tcp connection is failing and wanted to know why so I will remove the or we might have interest to know what was the tcp flags when drop occurs

Reply:
Hey,
I did also discuss this with @msherif1234 - I agree that it is quite arbitrary to show only the last reason, but there's also a tradeoff to do between debugability and complexity (and performance impact - all events will be stored), so it seems reasonable to me to only pick 1 reason; that said, I cannot tell if the last would be statistically more significant.. maybe the first would be more? Hard to say, perhaps yes. Any thought about that?
Keeping all flags was easier as it doesn't involve to re-shape the data model.
Perhaps to avoid confusion, the `drop_cause` should be renamed `last_drop_cause` ?

Reply:
I think keeping history of all drops for older transient drops will just add confusion without much value imo I am fine with rename if that will make it more clear

Reply:
I don't think we can predict, in general, which drop will be the most significant to debug any particular issue. It really depends on what kind of connection issue you're trying to debug; I *suspect* having the full drop history will turn out to be useful in the future, but I really don't have anything other than a feeling to back that up with, so having just the last reason is probably a reasonable start... :) Renaming the field so anyone looking at it can tell that it's in fact the last reason is a good idea, though!

To expand a bit on why I think OR'ing the flags field is a bad idea: the issue is basically that the flags are not exclusive. E.g., the first packet in a connection will have the SYN flag set, the second one will have SYN+ACK, and subsequent packets will have just the ACK packet set. So say you look at the data and you see the drop counter has a value of 2, and the flags field is set to SYN+ACK. Does that mean that the last packet dropped was a SYN+ACK packet, or was there an earlier SYN packet that got dropped, retransmitted, and later a single ACK got dropped? You have no way to tell because of the OR'ing. So better to keep just the last flags so you can unambiguously refer that to the drop reason; and if you need the full history of drops for a connection, add proper history tracking with separate events per drop.

Reply:
Well, we have a trick here:
```c
// Custom flags exported
#define SYN_ACK_FLAG 0x100
#define FIN_ACK_FLAG 0x200
#define RST_ACK_FLAG 0x400
```
:-)

But anyway, that's fine with me to keep latest flag only! We don't have any strong requirement about that